### PR TITLE
[minio] Add types for minio extensions

### PR DESCRIPTION
**Pull request by MaDKaTZe**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <https://docs.min.io/docs/javascript-client-api-reference#listObjectsV2WithMetadata>

### CODE REVIEW

File: types/minio/index.d.ts [hunk 1]
```diff
@@ -187,7 +191,12 @@ export class Client {
 
     // Other
     newPostPolicy(): PostPolicy;
-    setRequestOptions(otpions: AgentOptions): void;
+    setRequestOptions(options: AgentOptions): void;
+
+    // Minio extensions that aren't necessary present for Amazon S3 compatible storage servers
+    extensions: {
+        listObjectsV2WithMetadata(bucketName: string, prefix?: string, recursive?: boolean, startAfter?: string): BucketStream<BucketItemWithMetadata>;
+    };
```
Comment (OutdatedVersion) [inline types/minio/index.d.ts:199]: Are you familiar with what cases `extensions` are available? I've tried reading the documentation but couldn't find any notes on their "extension" system.

**Review by OutdatedVersion — COMMENTED** [review]
This tentatively looks fine. _If_ we can enhance the types to only expose `extensions` when the API is available (as an extension, I assume there is a precursor step?) that would be great.

File: types/minio/index.d.ts (hunk 1 above)
Comment (barinbritva) [reply types/minio/index.d.ts:199]: I found something. It's [in the doc](https://docs.min.io/docs/javascript-client-api-reference#listObjectsV2WithMetadata) below `listObjectsV2WithMetadata` method:

```
var stream = minioClient.extensions.listObjectsV2WithMetadata('mybucket','', true,'')
stream.on('data', function(obj) { console.log(obj) } )
stream.on('error', function(err) { console.log(err) } )
```

Comment (barinbritva) [reply types/minio/index.d.ts:199]: It seems it's some part of server configuration - https://docs.min.io/docs/minio-bucket-replication-guide.html
In heading MinIO Extension
🤔

Comment (MaDKaTZe) [reply types/minio/index.d.ts:199]: > Are you familiar with what cases `extensions` are available? I've tried reading the documentation but couldn't find any notes on their "extension" system.

From the client's perspective, this property (the *extensions*) is available as a lazy getter (as seen [here](https://github.com/minio/minio-js/blob/master/src/main/minio.js#L3047)). Actual extensions are defined in a separate file - [here](https://github.com/minio/minio-js/blob/master/src/main/extensions.js). No server checks are performed on the backend side to see if they are indeed supported. Under the hood, this method is using the S3 *listObjectsV2* API with an additional query parameter `metadata=true`.

As for the server configuration goes - I'm not aware if there are any specific ones, that must be set before using this functionality. Using a pre-built Docker image, this works just fine out of the box.

**Review by barinbritva — CHANGES_REQUESTED** [review]
@MaDKaTZe, I see there are 2 methods in `extensions.js`.  Can I ask you add second one also - `listObjectsV2WithMetadataQuery`?

**Comment by MaDKaTZe** [comment]
> @MaDKaTZe, I see there are 2 methods in `extensions.js`. Can I ask you add second one also - `listObjectsV2WithMetadataQuery`?

I could, yes. However, there is something that worries me a little - this method isn't mentioned anywhere in the library's documentation and from the looks of the source code. Also, it seems that the result returns a transformer function instead of a stream of listed objects as returned by `listObjectsV2WithMetadata` for example. The method doesn't seem to contain `__Arguments__` and ` __Return Value__` annotations, nor a full description of the return value, seemingly present for all public methods.  My assumption is that it shouldn't be really used outside the class, but the source base doesn't contain a clear indication of whether a method should be considered public or private so I could be wrong.

**Review by barinbritva — APPROVED** [review]
@MaDKaTZe yeah, you are right

**Comment by MaDKaTZe** [comment]
Ready to merge

**Pull request merged**